File: lib/infrastructure/routing/auth_redirect_policy.dart
import 'package:iamhere/feature/auth/service/auth_state.dart';

import 'app_routes.dart';

class AuthRedirectPolicy {
  const AuthRedirectPolicy();

  String? resolve({
    required AuthState? authState,
    required bool autoSendReady,
    required String matchedLocation,
    required Uri requestedUri,
  }) {
    final isPending = authState == AuthState.pending;
    final isInactive = authState == AuthState.inactive;
    final isAuthenticated = authState == AuthState.authenticated;

    if (!isAuthenticated && !isPending && !isInactive) {
      if (matchedLocation == AppRoutes.auth) return null;

      return Uri(
        path: AppRoutes.auth,
        queryParameters: {'redirect': requestedUri.toString()},
      ).toString();
    }

    if (isPending) {
      if (matchedLocation == AppRoutes.termsConsent) return null;
      return Uri(
        path: AppRoutes.termsConsent,
        queryParameters: {'redirect': requestedUri.toString()},
      ).toString();
    }

    if (isInactive) {
      if (matchedLocation == AppRoutes.auth) return null;
      return Uri(
        path: AppRoutes.auth,
        queryParameters: {'reason': 'inactive'},
      ).toString();
    }

    if (!autoSendReady) {
      if (matchedLocation == AppRoutes.userPermission ||
          matchedLocation == AppRoutes.locationPermissionGuide ||
          matchedLocation == AppRoutes.batteryOptimizationGuide) {
        return null;
      }

      return Uri(
        path: AppRoutes.userPermission,
        queryParameters: {'redirect': requestedUri.toString()},
      ).toString();
    }

    if (matchedLocation == AppRoutes.userPermission) {
      final redirect = requestedUri.queryParameters['redirect'];
      if (redirect != null && redirect.startsWith('/')) {
        return redirect;
      }

      return AppRoutes.geofence;
    }

    if (matchedLocation == AppRoutes.locationPermissionGuide ||
        matchedLocation == AppRoutes.batteryOptimizationGuide) {
      return null;
    }

    if (matchedLocation == AppRoutes.auth ||
        matchedLocation == AppRoutes.termsConsent) {
      final redirect = requestedUri.queryParameters['redirect'];
      if (redirect != null && redirect.startsWith('/')) {
        return redirect;
      }

      return AppRoutes.geofence;
    }

    return null;
  }
}

File: web/src/auth/auth-redirect-policy.ts
export type UserStatus = "pending" | "active" | "inactive" | null;

export interface AuthSnapshot {
  authenticated: boolean;
  userStatus: UserStatus;
}

const routes = {
  auth: "/auth",
  termsConsent: "/terms-consent",
  userPermission: "/user-permission",
  locationGuide: "/location-permission-guide",
  batteryGuide: "/battery-optimization-guide",
  geofence: "/geofence",
} as const;

const permissionGuideRoutes = new Set<string>([
  routes.userPermission,
  routes.locationGuide,
  routes.batteryGuide,
]);

export function resolveAuthRedirect({
  auth,
  autoSendReady,
  requestedUrl,
}: {
  auth: AuthSnapshot;
  autoSendReady: boolean;
  requestedUrl: string;
}): string | null {
  const requested = new URL(requestedUrl, "https://app.invalid");
  const matchedLocation = requested.pathname;
  const redirect = requested.pathname + requested.search + requested.hash;

  if (
    !auth.authenticated &&
    auth.userStatus !== "pending" &&
    auth.userStatus !== "inactive"
  ) {
    return matchedLocation === routes.auth
      ? null
      : withQuery(routes.auth, "redirect", redirect);
  }

  if (auth.userStatus === "pending") {
    return matchedLocation === routes.termsConsent ||
      matchedLocation.startsWith("/terms-detail/")
      ? null
      : withQuery(routes.termsConsent, "redirect", redirect);
  }

  if (auth.userStatus === "inactive") {
    return matchedLocation === routes.auth
      ? null
      : withQuery(routes.auth, "reason", "inactive");
  }

  if (!autoSendReady) {
    return permissionGuideRoutes.has(matchedLocation)
      ? null
      : withQuery(routes.userPermission, "redirect", redirect);
  }

  if (matchedLocation === routes.userPermission) {
    return (
      safeRedirect(requested.searchParams.get("redirect")) ?? routes.geofence
    );
  }

  if (
    matchedLocation === routes.locationGuide ||
    matchedLocation === routes.batteryGuide
  ) {
    return null;
  }

  if (
    matchedLocation === routes.auth ||
    matchedLocation === routes.termsConsent
  ) {
    return (
      safeRedirect(requested.searchParams.get("redirect")) ?? routes.geofence
    );
  }

  return null;
}

function withQuery(path: string, key: string, value: string) {
  const query = new URLSearchParams({ [key]: value });
  return `${path}?${query}`;
}

function safeRedirect(value: string | null) {
  return value?.startsWith("/") === true && !value.startsWith("//")
    ? value
    : null;
}
